pkg/controllers/kube-apiserver.go

```diff
@@ -16,130 +16,193 @@ limitations under the License.
 package controllers
 
 import (
+	"bytes"
 	"context"
+	"encoding/json"
 	"fmt"
+	"io"
 	"io/ioutil"
 	"net"
 	"os"
 	"path/filepath"
+	"strconv"
 	"time"
 
-	"github.com/openshift/microshift/pkg/config"
-
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/serializer"
-	utilerrors "k8s.io/apimachinery/pkg/util/errors"
-	utilnet "k8s.io/apimachinery/pkg/util/net"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
 	"k8s.io/klog/v2"
 	kubeapiserver "k8s.io/kubernetes/cmd/kube-apiserver/app"
 	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+
+	configv1 "github.com/openshift/api/config/v1"
+	kubecontrolplanev1 "github.com/openshift/api/kubecontrolplane/v1"
+	"github.com/openshift/library-go/pkg/crypto"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/microshift/pkg/assets"
+	"github.com/openshift/microshift/pkg/config"
 )
 
 const (
 	kubeAPIStartupTimeout = 60
 )
 
+var baseKubeAPIServerConfigs = [][]byte{
+	// todo: a nice way to generate the baseline kas config for microshift
+	assets.MustAsset("assets/components/kube-apiserver/defaultconfig.yaml"),
+	assets.MustAsset("assets/components/kube-apiserver/config-overrides.yaml"),
+}
+
+var fixedTLSProfile *configv1.TLSProfileSpec
+
+func init() {
+	var ok bool
+	fixedTLSProfile, ok = configv1.TLSProfiles[configv1.TLSProfileIntermediateType]
+	if !ok {
+		panic("lookup of intermediate tls profile failed")
+	}
+}
+
 type KubeAPIServer struct {
+	kasConfigBytes []byte
+	verbosity      int
+	configureErr   error // todo: report configuration errors immediately
+
 	serverOptions *options.ServerRunOptions
 
 	kubeconfig string
 }
 
 func NewKubeAPIServer(cfg *config.MicroshiftConfig) *KubeAPIServer {
 	s := &KubeAPIServer{}
-	s.configure(cfg)
+	if err := s.configure(cfg); err != nil {
+		s.configureErr = err
+	}
 	return s
 }
 
 func (s *KubeAPIServer) Name() string           { return "kube-apiserver" }
 func (s *KubeAPIServer) Dependencies() []string { return []string{"etcd"} }
 
-func (s *KubeAPIServer) configure(cfg *config.MicroshiftConfig) {
+func (s *KubeAPIServer) configure(cfg *config.MicroshiftConfig) error {
+	s.kubeconfig = filepath.Join(cfg.DataDir, "resources", "kube-apiserver", "kubeconfig")
+	s.verbosity = cfg.LogVLevel
+
 	caCertFile := filepath.Join(cfg.DataDir, "certs", "ca-bundle", "ca-bundle.crt")
-	// certDir := filepath.Join(cfg.DataDir, "certs", s.Name())
-	// dataDir := filepath.Join(cfg.DataDir, s.Name())
 
 	if err := s.configureAuditPolicy(cfg); err != nil {
-		klog.Fatalf("Failed to configure kube-apiserver audit policy %v", err)
+		return fmt.Errorf("Failed to configure kube-apiserver audit policy: %w", err)
```

Comment:
Why not using klog? It's our logging system for the entire project.

Reply:
Fatalf is equivalent to log + `os.Exit(1)`. It's preferable to follow normal control flow here, return an error, and let the service manager gracefully stop all of the components before the process exits.